File: backend/app/schemas/plan.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import ProcessingStatus
from app.schemas.common import PlanTaskStatus


class PlanBase(BaseModel):
    title: str = Field(description="Plan title")
    content: str | None = Field(default=None, description="Markdown content describing the plan")


class PlanCreate(PlanBase):
    parent_task_id: UUID | None = Field(
        default=None, description="Task ID if this plan was spawned by a complex task"
    )


class PlanUpdate(BaseModel):
    title: str | None = Field(default=None, description="Plan title")
    content: str | None = Field(default=None, description="Markdown content")


class Plan(PlanBase):
    id: UUID = Field(description="Plan unique identifier")
    project_id: UUID = Field(description="Parent project ID")
    status: PlanTaskStatus = Field(description="Current plan status")
    parent_task_id: UUID | None = Field(
        default=None, description="Task ID if this plan was spawned by a complex task"
    )
    version: int = Field(default=1, description="Plan version number")
    processing_status: ProcessingStatus | None = Field(
        default=None, description="Status of AI content generation"
    )
    processing_error: str | None = Field(
        default=None, description="Error message if generation failed"
    )
    created_by: UUID | None = Field(default=None, description="User or system that created")
    created_at: datetime = Field(description="Creation timestamp")
    updated_at: datetime | None = Field(default=None, description="Last update timestamp")

    model_config = {"from_attributes": True}


class PlanWithDetails(Plan):
    tasks: list["TaskSummary"] = Field(default_factory=list, description="Tasks spawned by plan")
    reviews: list["ReviewSummary"] = Field(default_factory=list, description="Reviews on this plan")
    threads: list["ThreadSummary"] = Field(
        default_factory=list, description="Comment threads on this plan"
    )


class TaskSummary(BaseModel):
    id: UUID
    title: str
    status: PlanTaskStatus


class ReviewSummary(BaseModel):
    id: UUID
    reviewer_id: UUID
    decision: str
    created_at: datetime


class ThreadSummary(BaseModel):
    id: UUID
    status: str
    comment_count: int


class PlanGenerateRequest(BaseModel):
    """Request to generate plan content using Claude."""

    context: str | None = Field(
        default=None,
        description="Additional context to provide to Claude for plan generation",
    )


class PlanGenerationStatus(BaseModel):
    """Response showing the current generation status of a plan."""

    plan_id: UUID = Field(description="Plan ID")
    processing_status: ProcessingStatus | None = Field(
        description="Current processing status"
    )
    processing_error: str | None = Field(
        default=None, description="Error message if generation failed"
    )
    content: str | None = Field(
        default=None, description="Generated content if completed"
    )


class SpawnTasksRequest(BaseModel):
    """Request to spawn tasks from an approved plan."""

    pass  # No additional parameters needed; plan content is used


class SpawnedTaskSummary(BaseModel):
    """Summary of a task spawned from a plan."""

    id: UUID = Field(description="Task ID")
    title: str = Field(description="Task title")
    description: str | None = Field(default=None, description="Task description")
    blocked_by: list[UUID] = Field(
        default_factory=list, description="IDs of tasks that block this task"
    )


class SpawnTasksResponse(BaseModel):
    """Response from spawning tasks from a plan."""

    plan_id: UUID = Field(description="Plan ID that spawned the tasks")
    tasks_created: int = Field(description="Number of tasks created")
    tasks: list[SpawnedTaskSummary] = Field(description="List of spawned tasks")


class SpawnTasksStatus(BaseModel):
    """Status response for task spawning progress."""

    plan_id: UUID = Field(description="Plan ID")
    processing_status: ProcessingStatus | None = Field(
        description="Current processing status"
    )
    processing_error: str | None = Field(
        default=None, description="Error message if spawning failed"
    )
    tasks_created: int | None = Field(
        default=None, description="Number of tasks created (if completed)"
    )


PlanWithDetails.model_rebuild()

File: frontend/src/types/plan.ts
// Plan status enum - matches backend PlanTaskStatus
export type PlanStatus =
  | 'backlog'
  | 'blocked'
  | 'coding'
  | 'review'
  | 'approved'
  | 'cicd'
  | 'merged'
  | 'closed'

// Enum version for easier usage with switch statements
export enum PlanTaskStatus {
  BACKLOG = 'backlog',
  BLOCKED = 'blocked',
  CODING = 'coding',
  REVIEW = 'review',
  APPROVED = 'approved',
  CICD = 'cicd',
  MERGED = 'merged',
  CLOSED = 'closed',
}

// Processing status for AI-generated content
export type ProcessingStatusType = 'pending' | 'generating' | 'completed' | 'failed'

// Enum version for easier usage
export enum ProcessingStatus {
  PENDING = 'pending',
  GENERATING = 'generating',
  COMPLETED = 'completed',
  FAILED = 'failed',
}

// Base plan interface
export interface Plan {
  id: string
  project_id: string
  title: string
  content: string | null
  status: PlanStatus
  parent_task_id: string | null
  version: number
  processing_status: ProcessingStatus | null
  processing_error: string | null
  created_by: string | null
  created_at: string
  updated_at: string | null
}

// Summary interfaces for related entities
export interface TaskSummary {
  id: string
  title: string
  status: PlanStatus
}

export interface ReviewSummary {
  id: string
  reviewer_id: string
  decision: string
  created_at: string
}

export interface ThreadSummary {
  id: string
  status: string
  comment_count: number
}

// Plan with all related entities
export interface PlanWithDetails extends Plan {
  tasks: TaskSummary[]
  reviews: ReviewSummary[]
  threads: ThreadSummary[]
}

// Request to create a plan
export interface PlanCreate {
  title: string
  content?: string | null
  parent_task_id?: string | null
}

// Request to update a plan
export interface PlanUpdate {
  title?: string
  content?: string | null
}

// Request to generate plan content
export interface PlanGenerateRequest {
  context?: string | null
}

// Response from generation status endpoint
export interface PlanGenerationStatus {
  plan_id: string
  processing_status: ProcessingStatus | null
  processing_error: string | null
  content: string | null
}

// Request to spawn tasks from a plan
export interface SpawnTasksRequest {
  // No additional parameters needed
}

// Summary of a spawned task
export interface SpawnedTaskSummary {
  id: string
  title: string
  description: string | null
  blocked_by: string[]
}

// Response from spawning tasks
export interface SpawnTasksResponse {
  plan_id: string
  tasks_created: number
  tasks: SpawnedTaskSummary[]
}

// Status response for task spawning
export interface SpawnTasksStatus {
  plan_id: string
  processing_status: ProcessingStatus | null
  processing_error: string | null
  tasks_created: number | null
}

// For the status filter dropdown
export const PLAN_STATUS_OPTIONS: { value: PlanStatus | ''; label: string }[] = [
  { value: '', label: 'All Statuses' },
  { value: 'backlog', label: 'Backlog' },
  { value: 'blocked', label: 'Blocked' },
  { value: 'coding', label: 'Coding' },
  { value: 'review', label: 'Review' },
  { value: 'approved', label: 'Approved' },
  { value: 'cicd', label: 'CI/CD' },
  { value: 'merged', label: 'Merged' },
  { value: 'closed', label: 'Closed' },
]

// Status display configuration
export const PLAN_STATUS_CONFIG: Record<
  PlanStatus,
  { label: string; color: string; bgColor: string }
> = {
  backlog: { label: 'Backlog', color: '#6b7280', bgColor: '#f3f4f6' },
  blocked: { label: 'Blocked', color: '#dc2626', bgColor: '#fef2f2' },
  coding: { label: 'Coding', color: '#2563eb', bgColor: '#eff6ff' },
  review: { label: 'Review', color: '#d97706', bgColor: '#fffbeb' },
  approved: { label: 'Approved', color: '#059669', bgColor: '#ecfdf5' },
  cicd: { label: 'CI/CD', color: '#7c3aed', bgColor: '#f5f3ff' },
  merged: { label: 'Merged', color: '#0891b2', bgColor: '#ecfeff' },
  closed: { label: 'Closed', color: '#374151', bgColor: '#f9fafb' },
}

// Processing status display configuration
export const PROCESSING_STATUS_CONFIG: Record<
  ProcessingStatus,
  { label: string; color: string; bgColor: string }
> = {
  [ProcessingStatus.PENDING]: { label: 'Pending', color: '#6b7280', bgColor: '#f3f4f6' },
  [ProcessingStatus.GENERATING]: { label: 'Generating...', color: '#2563eb', bgColor: '#eff6ff' },
  [ProcessingStatus.COMPLETED]: { label: 'Generated', color: '#059669', bgColor: '#ecfdf5' },
  [ProcessingStatus.FAILED]: { label: 'Failed', color: '#dc2626', bgColor: '#fef2f2' },
}

// Helper type for display purposes (enum-based labels)
export const PlanTaskStatusLabels: Record<PlanTaskStatus, string> = {
  [PlanTaskStatus.BACKLOG]: 'Backlog',
  [PlanTaskStatus.BLOCKED]: 'Blocked',
  [PlanTaskStatus.CODING]: 'Coding',
  [PlanTaskStatus.REVIEW]: 'Review',
  [PlanTaskStatus.APPROVED]: 'Approved',
  [PlanTaskStatus.CICD]: 'CI/CD',
  [PlanTaskStatus.MERGED]: 'Merged',
  [PlanTaskStatus.CLOSED]: 'Closed',
}

export const ProcessingStatusLabels: Record<ProcessingStatus, string> = {
  [ProcessingStatus.PENDING]: 'Pending',
  [ProcessingStatus.GENERATING]: 'Generating...',
  [ProcessingStatus.COMPLETED]: 'Completed',
  [ProcessingStatus.FAILED]: 'Failed',
}
